### ISSUE TITLE
CSS history knocking / leak fixed?

### ISSUE THREAD

**Issue by mikeg-de** [samyk/evercookie#115]
Obtaining the computed color of visited links seems to be fixed back in 2010: http://stackoverflow.com/questions/5394099/detect-visited-link-in-chrome

Visited colors may be visible but the standard value is retrieved / provided by the browser.

**Comment by samyk** [comment]
True. It is still possibly exploitable and I'm open to suggestions or updates. HSTS is also worth looking into, although there have been some fixes there too: https://github.com/diracdeltas/sniffly

**Comment by mikeg-de** [comment]
I have tried it with:
- width, height
- font-size, line height, letter-spacing, font-family
- margin, padding (pos. and neg.)
- position
- display
- selectors like + or ~ with a second control link
- nested spans with the above declarations
- color rgba, hsla
- border or background image
- border-clip
- transition

but no effect could be accesed by getComputedStyle as stated on [developer.mozilla.org](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle#Notes).

Though I had the idea of using the HTML5 Resource Timing API. By assuming someone has visited a particular website, i.e. a company wide login page, it's resources or connection information should be cached. Attaching the resource to the DOM and checking the resource timing metrics would reveal if the user has visited the website.

The limitations I could not probe completely at present but I think the major limitations are:

- Cache limitations like no-cache
- Missing timing-allow-origin response response header
- Backwards compatibility [canisue.com](http://caniuse.com/#feat=resource-timing) pointing out Safari as the bottle neck

Check it out: [User identification with HTML5 Resource API](https://github.com/mikeg-de/User-identification-with-HTML5-Resource-API/wiki)

In my prove of concept it seems to work:
![1st-load-no-cache](https://cloud.githubusercontent.com/assets/604842/16758641/208e08ba-4813-11e6-987d-8784e1d294c3.jpg)
![2nd-load-with-deleted-cookie](https://cloud.githubusercontent.com/assets/604842/16758642/2090d04a-4813-11e6-8847-7fca03ba0a49.jpg)

**Comment by samyk** [comment]
Good testing. Have you tried seeing if you could glean information by using the visited css attribute to adjust the size of one object, but then obtaining information on a *different* object on the page which would have been affected? For example, if something is larger because it was visited, the object next to it should be in a different position on the page (or perhaps wrap).

Another thing I'd suggest investigating, though this is on a per-page basis, is if a resource behaves differently if a cookie is already set (user already visited or logged into the site), using that resource in a script tag or CSS tag. If you can use the remote resource as a CSS object style (even if it's not entirely valid), and some of the page is different based on whether the user has visited or logged into it before, then you may be able to glean if they've been to that site or more likely are logged in.

**Comment by samyk** [comment]
Also worth investigating:
https://lcamtuf.blogspot.com/2016/08/css-mix-blend-mode-is-bad-for-keeping.html
